### PR TITLE
New web GUI based on bootstrap.js

### PR DESCRIPTION
**Pull request by exilaus**
### Requirements

- Needs WIFISUPPORT only.
- Replaces current `data\www` files.
- [v1.1 for marlin.zip](https://github.com/MarlinFirmware/Marlin/files/5400097/v1.1.for.marlin.zip)

### Description
Advanced web socket for default Marlin websocket (no esplib no esp3d).

Unique part no work are upload file, have issue to sync with `M28` G-code on web socket, feel need improve Marlin to have direct access of SD (for cpu like ESP32).

Tested on E4d@box board.

Live Demo: https://e4d.atbox.tech/Gui/?ip=10.10.10.10 (upload version in this post no have any reference to e4d@box)

### Benefits
improve web gui and no use external projects

### Issues
Sync issue on upload file via `M28` G-code. Need help to improve default Marlin web implementation.

### CODE REVIEW

**Comment by thinkyhead** [comment]
Thanks! I did a large merge to bring this up to date, and since this PR was originally based on code from ancient times you may want to recheck to make sure it still functions with the latest codebase.

File: data/www/index.html [hunk 1]
```diff
@@ -1,37 +1,750 @@
...
+  <!-- MODAL: About window ========================================================================================= -->
+  <div class="modal fade" id="modal-info" tabindex="-1" role="dialog" aria-hidden="true">
+    <div class="modal-dialog modal-popup" role="document">
+      <div class="modal-content">
+        <div class="modal-header bg-secondary text-light m-0 px-2 py-1">
+          <span class="h5 font-italic mt-1"><i class="icon icon-info mr-2"></i>Info...</span>
+          <button type="button" class="close" data-dismiss="modal" aria-label="Close"><span aria-hidden="true">&times;</span></button>
+        </div>
+        <div class="modal-body p-2 text-left">
+          <span class="h3">E4d@box Marin WebUI 1.0</span>
```
Comment (thisiskeithb) [inline data/www/index.html]: Small typo here. Marin should be Marlin.

**Comment by exilaus** [comment]
> I did a large merge to bring this up to date, and since this PR was originally based on code from ancient times you may want to recheck to make sure it still functions with the latest codebase.

Sure, that developed  on 2.0.6  bugfix...anyway give me 2 day recheck all with last bugfix.  

Next steps are:
- implement another method for upload file from web
- where possible ,like in esp32 , move webserver/websocket process on core1 for dedicate core0 to Marlin .

File: data/www/index.html (hunk 1 above)
Comment (exilaus) [reply data/www/index.html]: ops! ok correct that in next review. many thanks.

**Comment by thisiskeithb** [comment]
Can the main title/header be simplified to just “Marlin WebUI” or use the printer’s name like “3D Printer WebUI” (since there’s already a Marlin logo).

The “E4d@box” credit would still be listed on the info page, so it wouldn’t be removed.

**Comment by exilaus** [comment]
little issue guys, i just upload last bugfix with new GUI and old GUI same result http server on wesocket feel down.. start to dig for undestand why.

image

**Comment by exilaus** [comment]
found why no see websocket never start MYSERIAL1 from marlincore.cpp and never start.
now force start to use MYSERIAL1  and now web socket open but feel other changed continue to investigate.

image

**Comment by luc-github** [comment]
looks like now NUM_SERIAL  is not set like before in ESP32 Hal (https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.0.x/Marlin/src/HAL/ESP32/HAL.h#L56-L62) but in main Hal (https://github.com/MarlinFirmware/Marlin/blob/bugfix-2.0.x/Marlin/src/HAL/HAL.h#L28-L32) and so not set anymore for ESP32 and wifi bridge
```
#ifdef SERIAL_PORT_2
  #define NUM_SERIAL 2
#else
  #define NUM_SERIAL 1
#endif
```
since : 
https://github.com/MarlinFirmware/Marlin/commit/4b928b2da8ae26bd4a949debe470a914e23545ca#diff-b433648b73ac70dda9c5db91bedfecab3660ca687743277c040cd870fa60c5ec

**Comment by exilaus** [comment]
Luc you are right ... 
But resolved without touching code  setting configuration.h  Need enable  SERIAL_PORT_2 -1  and websocket working fine with that parameter.

**Comment by vivian-ng** [comment]
While this is being added, how about a minor fix to `HAL.cpp` for ESP32? Haven't had time to look into it, but it affects the stepper control on ESP32-based boards that do not implement I2S stepper stream.

Change:
```
void HAL_init() { i2s_init(); }
```
to:
```
void HAL_init(void) {
#if ENABLED(I2S_STEPPER_STREAM)
    i2s_init();
#endif
}
```

Linked to this [issue](https://github.com/maplerainresearch/MRR_ESPA/issues/5#issuecomment-670544471).

**Comment by exilaus** [comment]
> While this is being added, how about a minor fix to `HAL.cpp` for ESP32? Haven't had time to look into it, but it affects the stepper control on ESP32-based boards that do not implement I2S stepper stream.

Vivian-ng please can contact me and check and fix thogheter that part ok? regards Mauro. 
ok checked post linked too verify and put in pull that fix too.

File: data/www/index-ie.html [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+    <meta name="Owner" Content="E4d@box" />
+    <meta name="Author" Content="MandrakeDesign" />
+    <meta name="google" content="notranslate" />
+    <meta http-equiv="Content-Language" content="en_GB" />
+    <title>E4d@box Marlin WEBUI</title>
+    <link rel="shortcut icon" href="#" />
+  </head>
+  <body>
+    <div style="padding-top:100px; text-align:center">
+      <h1>Sorry, but MS Internet Explorer browser is not supported by this web application.<br>Please upgrade to Edge or use another browser</h1>
```
Comment (thinkyhead) [inline data/www/index-ie.html:17]: What prevents IE support? What percentage of Windows users now uses IE as their default browser?

**Comment by thinkyhead** [comment]
Let's continue to refine and improve. It would be good to see the build info, machine name, etc., in the title, meta info, and on the page.

The word on the street is that Duet3D / RepRapFirmware dropped **bootstrap.js** for their web interface in favor of [vue.js](https://vuejs.org) because they were hitting limitations.

**Comment by rhapsodyv** [comment]
> Let's continue to refine and improve. It would be good to see the build info, machine name, etc., in the title, meta info, and on the page.
> 
> The word on the street is that Duet3D / RepRapFirmware dropped **bootstrap.js** for their web interface in favor of [vue.js](https://vuejs.org) because they were hitting limitations.

I use react in my startup. But, for Marlin, react, vue or angular, could add complexity for the users, we will need a JS bundler to build and process the Javascript. Bootstrap+html just run as it is... 

I'm not against it, just a warning point.

**Comment by thinkyhead** [comment]
This is a good starting-point and should be refined, quite a lot. Modern displays have a lot of space, and this seems to be optimized for a small Raspberry Pi display. I would like to see a clean aesthetic similar to that of the OctoPrint default theme applied to all of our interfaces.

**Comment by darkberg** [comment]
> This is a good starting-point and should be refined, quite a lot. Modern displays have a lot of space, and this seems to be optimized for a small Raspberry Pi display. I would like to see a clean aesthetic similar to that of the OctoPrint default theme applied to all of our interfaces.

Actually, it was made  for mobile phone control, with minimal data usage. Of course, it's a starting point, to test / control E4d@box board.  Any updates are / will be more than welcomed.

**Pull request merged**